### DISCUSSION TITLE
button HOLD action only happen once

### DISCUSSION THREAD

**Discussion by ChrisVonFrosta** [arendst/Tasmota#22524] [Q&A]
i would like to use long press aka "HOLD" for dimming or colormixing in berry.
I expacted that HOLD mesage repeats in interval from SetOption62.
But they occures only once per button press.
Is the an option i overlooked?

**Comment by barbudor** [comment]
Have a look to Switchmode 11 and 12 in https://tasmota.github.io/docs/Buttons-and-Switches/#switchmode

The rules cookbook has an example on how to use that that you can probably transpose to Berry
https://tasmota.github.io/docs/Rules/#control-a-dimmer-with-one-switch

**Reply by sfromis** [reply]
FTR, `SetOption62` has nothing to do with repeating, is instead about using (or not) the MQTT retain flag for messages already published.

**Reply by ChrisVonFrosta** [reply]
>Have a look to Switchmode 11 and 12 ...

this has to been set for each button individually?

also I see, that `switchmode` is only for switches, not for buttons?!
![grafik](https://github.com/user-attachments/assets/c63b1756-bf22-4132-acbf-1b4c39965cf8)


`mcp23x.dat`:
```
{"NAME":"MCP23017 A=B1-8, B=B9-16, C=B17-24, D=B25-32","GPIO":[32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63]}
```

i tried `switchmode1 11` but i got only one "HOLD" message per "push&hold" event, Don't matter how long I keep the button presssed.
```
10:29:25.730 CMD: setoption32
10:29:25.733 RSL: RESULT = {"SetOption32":5}
10:29:39.713 CMD: switchmode1
10:29:39.716 RSL: RESULT = {"SwitchMode1":11}
```

**Reply by sfromis** [reply]
It can't be for a `Button` pin, only for individual `Switch` pins. You need to be very specific about not just talking about a physical thing, as Tasmota can handle such stuff in quite different ways, configured as Button vs Switch inputs.

Configuring as "Button" is only reasonable for a momentary push, while configuring as "Switch" can both work for locking toggles or momentary push switches, due to the many behavior variations available with `SwitchMode` - but still not with all the features specific to Button configs.

**Reply by barbudor** [reply]
And to your question, yes it has to be applies to each `switch` individually like
```
switchmode4 11
```
To set `Switch 4` to mode 11
This is a permanent configuration (saved in flash settings) Tha you have to enter only once in the console

**Reply by ChrisVonFrosta** [reply]
with other words: i cannot configure the harware pushbuttons on MCP23017 as buttons AND get the repeated HOLD message?
i have to configure them as switches in `mcp27x.dat, i haven't?

**Reply by barbudor** [reply]
Yes

**Reply by ChrisVonFrosta** [reply]
thx. but the amount of switches is limited to 28 instead of 32?

**Reply by sfromis** [reply]
From what I can see (no testing, but matches what I see in the dropdowns), you can have up to 32 switches, but only 28 `SwitchText` strings.

**Comment by arendst** [comment]
32 switches only in latest dev v14.3.0.7

**Reply by ChrisVonFrosta** [reply]
not needed yet. i was just wondering.

but it this correctly?
![grafik](https://github.com/user-attachments/assets/fc9299e8-08b2-480a-b8bf-8ec47f9cf625)

pull-UPs for both logics?

**Reply by barbudor** [reply]
Because ESP8266 only has internal pull ups
Button_i is usabled only with a push-to-break
Have you seen the diagrams at the beginning of this section ? https://tasmota.github.io/docs/Buttons-and-Switches/#button

**Comment by ChrisVonFrosta** [comment]
i am sorry, but i don't get it with multi press AND repeated long press / HOLD function.

now i tried with switches in `mcp23x.dat` and switchmode1 12 but i got "Power_INV, Power_CLEAR, POWER_100", which i didn't find information about. With Buttons, it was close to perfect. Only the reapeated HOLD/long push was missed.

now:
```
21:21:24.539 CMD: switchmode1
21:21:24.542 RSL: RESULT = {"SwitchMode1":12}
21:21:30.124 CMD: setoption32
21:21:30.128 RSL: RESULT = {"SetOption32":10}
21:22:04.600 CMD: setoption114
21:22:04.603 RSL: RESULT = {"SetOption114":"ON"}
21:22:52.507 CMD: setoption1
21:22:52.509 RSL: RESULT = {"SetOption1":"ON"}
21:22:59.664 CMD: setoption11
21:22:59.666 RSL: RESULT = {"SetOption11":"ON"}
21:23:02.888 CMD: setoption13
21:23:02.892 RSL: RESULT = {"SetOption13":"OFF"}
#one double press#
21:23:06.268 RSL: RESULT = {"Switch1":{"Action":"TOGGLE"}}
21:23:06.367 RSL: RESULT = {"Switch1":{"Action":"POWER_INV"}}
21:23:06.567 RSL: RESULT = {"Switch1":{"Action":"POWER_100"}}
21:23:07.567 RSL: RESULT = {"Switch1":{"Action":"POWER_CLEAR"}}
```

`mcp27x.dat`
```
{"NAME":"MCP23017 A=B1-8, B=B9-16, C=B17-24, D=B25-32","GPIO":[160,161,162,163,164,165,166,167,168,169,170,171,172,173,174,175,176,177,178,179,180,181,182,183,184,185,186,187]}
```

**Reply by ChrisVonFrosta** [reply]
maybe i have to stay without builtin multipress engine and adapt my multipress berry parsing from PCF8574 to the MCP23017. Than i can go with  Buttons or Switchmode 1.

Or i try to go with "HOLD" for starting a long Press action and "RELEASE" to stop it
